lib/httparrot/p_handler.ex

```diff
@@ -57,4 +58,70 @@ defmodule HTTParrot.PHandler do
   defp response(info, body) do
     info ++ body |> JSEX.encode!
   end
+
+  def post_multipart(req, _state) do
+    {:ok, parts, req} = handle_multipart(req)
+
+    filter = fn({type, _name, _body}, type_atom) -> type == type_atom end
+    reducer = fn({_type, name, body}, acc) -> acc ++ [{name, body}] end
+
+    # the other post handlers return a list with a single empty tuple if
+    # there's no data for forms, so let's match that behavior...
+    normalize = fn(parts) -> if parts == [], do: [{}], else: parts end
+
+    file_parts = Enum.filter(parts, &(filter.(&1, :file)))
+      |> Enum.reduce([], &(reducer.(&1, &2)))
+      |> normalize.()
+
+    form_parts = Enum.filter(parts, &(filter.(&1, :form)))
+      |> Enum.reduce([], &(reducer.(&1, &2)))
+      |> normalize.()
+
+    post(req, [form: form_parts, files: file_parts, data: "", json: nil])
+  end
+
+  defp handle_multipart(req, parts \\ []) do
+    case :cowboy_req.part(req) do
+      {:done, req} -> {:ok, parts, req}
+      {:ok, headers, req} ->
+        content_disposition = List.keyfind(headers, "content-disposition", 0)
+        if content_disposition != nil do
+          case parse_content_disposition_header(content_disposition) do
+            %{:type => "form-data", "name" => name, "filename" => _filename} ->
+              {:ok, file, req} = handle_multipart_body(req)
+              handle_multipart(req, parts ++ [{:file, name, file}])
+            %{:type => "form-data", "name" => name} ->
+              {:ok, form_part, req} = handle_multipart_body(req)
+              handle_multipart(req, parts ++ [{:form, name, form_part}])
+            _ ->
+              {:ok, parts, req}
+          end
+        else
+          {:ok, parts, req}
+        end
+    end
+  end
+
+  defp handle_multipart_body(req, parts \\ []) do
+    case :cowboy_req.part_body(req) do
+      {:ok, data, req} ->
+        {:ok, Enum.join(parts ++ [data]), req}
+      {:more, data, req} ->
+        handle_multipart_body(req, parts ++ [data])
+    end
+  end
+
+  defp parse_content_disposition_header(header) do
+    parts = elem(header, 1) |> String.split(";")
+
+    type = Enum.at(parts, 0)
+    parts = Enum.drop(parts, 1)
+
+    Enum.reduce(parts, %{:type => type}, fn part, acc ->
```

Comment:
I think you can use for comprehension here to be more concise:

```elixir
    for part <- parts, into: %{type: type} do
      [key, value] = String.split(part, "=")
      key = String.strip(key)
      value = String.strip(value) |> String.replace("\"", "")
      {key, value}
    end
```

What do you think?